Support stateless multipart

jchavarri opened the pull request:
Fixes #39.

Adds the necessary types and functions to perform multipart uploads on a stateless environment.

The first commit 9793d10620e87bd7bf4de86316cd961436b855d4 adds a function `complete_stateless` that takes in the `parts` as argument, and rewrite `complete` in its terms.

The second commit 7a34222cb38294073d7da5ab310697e428314268 allows to pass a `custom_query` argument to `make_presigned_url` so that urls with `partNumber` and `uploadId` in its query params can be presigned.

Please let me know if anything has to be implemented differently, or any additional modifications are needed, thanks!

jchavarri committed:
support stateless multipart

jchavarri committed:
add custom query

andersfugmann commented on aws-s3/authorization.ml:
```diff
@@ -135,9 +135,12 @@ let make_presigned_url ?(scheme=`Https) ?host ?port ~credentials ~date ~region ~
       | Some token -> ("X-Amz-Security-Token", token) :: base
+    in
+    (* Merge custom query parameters with AWS parameters *)
+    custom_query @ base_with_token
```
Copilot, Do you have a link to the documentation for this requirement?

andersfugmann replied:
This is already done on line 39.

andersfugmann commented on aws-s3/s3.mli:
```diff
@@ -222,9 +222,20 @@ module Make(Io : Types.Io) : sig
+(** Complete a multipart upload with explicit parameters.
+    This is useful for stateless workflows where upload metadata is stored in a database.
+    The returned string is an opaque identifier used as etag.
+    the etag return is _NOT_ the md5 *)
```
Could the `Multipart.t` not be used here as it is in `complete` so `complete` and `complete_stateless` have identical signatures?

jchavarri replied:
> Could the `Multipart.t` not be used here as it is in `complete` so `complete` and `complete_stateless` have identical signatures?

I assume you meant "Could the Multipart.t **be used** here so `complete` and `complete_stateless` have identical signatures?".

If that's the case, I'm not sure we can do this as the interface exists now. The whole point of `complete_stateless` is to support stateless workflows where upload metadata is stored in a database. In such workflows, you typically don't have a `Multipart_upload.t` object in memory. You only have the individual pieces of data (bucket, key, upload_id, parts list) that you've stored and retrieved from your database.

If we want to do this change, then we have to either:
- make `Multipart.t` a concrete (i.e. non-abstract) record type, so consumers can create values from its individual parts
- keep it abstract, but add a `make` function to create values

Which way do you prefer? Or maybe there are some alternatives that I'm not seeing?

andersfugmann replied:
Oh - thanks for clarifying. I misread the intent here. I think it would be much cleaner to just add a `make` function to instantiate `Multipart.t` which can then be passed to `Multipart.complete`. Seems less invasive and cleaner API.

jchavarri replied:
Done in https://github.com/andersfugmann/aws-s3/pull/40/commits/8e477dba73f06924c782ac0069660c054758ec8f.

andersfugmann reviewed (commented):
Thanks for the PR. Will look at this in detail over the weekend.

andersfugmann commented on aws-s3/authorization.ml:
```diff
@@ -135,9 +135,12 @@ let make_presigned_url ?(scheme=`Https) ?host ?port ~credentials ~date ~region ~
       | Some token -> ("X-Amz-Security-Token", token) :: base
+    in
+    (* Merge custom query parameters with AWS parameters *)
+    custom_query @ base_with_token
```
Challenged reading Ocaml code I gather :-)

andersfugmann commented on aws-s3/authorization.ml:
```diff
@@ -103,7 +103,7 @@ let make_auth_header ~credentials ~scope ~signed_headers ~signature =
     signature
 
-let make_presigned_url ?(scheme=`Https) ?host ?port ~credentials ~date ~region ~path ~bucket ~verb ~duration () =
+let make_presigned_url ?(scheme=`Https) ?host ?port ?(custom_query=[]) ~credentials ~date ~region ~path ~bucket ~verb ~duration () =
```
As `custom_query` argument a list of additional query parameters, I suggest changing the name to `query` to match the naming convention for the other functions in the module.

jchavarri replied:
done in https://github.com/andersfugmann/aws-s3/pull/40/commits/8ae2f327612c6953f5368d37a7e3322bda51c3b7.

andersfugmann commented on aws-s3/authorization.ml:
```diff
@@ -135,9 +135,12 @@ let make_presigned_url ?(scheme=`Https) ?host ?port ~credentials ~date ~region ~
       | Some token -> ("X-Amz-Security-Token", token) :: base
+    in
+    (* Merge custom query parameters with AWS parameters *)
+    custom_query @ base_with_token
```
Looking at the [AWS documentation on query signing](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html) for signing query parameters, it says that the parameter names should be sorted. However, Copilot seems quite adamant that values should be sorted if the keys are identical.

While I cannot find any direct documentation pointing to this, I think it's worth making this change. It should be simple to move the sorting from line 39 to line 68 and just sort on the generated strings.

jchavarri replied:
Done in https://github.com/andersfugmann/aws-s3/pull/40/commits/222013efc303ded64948c4c59be78b28e9f2b15c.

Found some [S3 docs](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html) where they mention it, so Copilot was right (this time 😅 ):

> `CanonicalQueryString` specifies the URI-encoded query string parameters. You URI-encode name and values individually. You must also sort the parameters in the canonical query string alphabetically by key name. The sorting occurs after encoding.

jchavarri committed:
rename custom_query to query

jchavarri committed:
sort headers after encoding

from s3 docs:
https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
> CanonicalQueryString specifies the URI-encoded
> query string parameters. You URI-encode name and
> values individually. You must also sort the parameters
> in the canonical query string alphabetically by key
> name. The sorting occurs after encoding.

jchavarri committed:
add Multipart_upload.make and fuse complete_stateless and complete

jchavarri committed:
cleanup test description

jchavarri commented:
Thanks for the fast review. I think I tackled all points discussed.

andersfugmann reviewed (approved):
LGTM. Thanks for the PR and quick turnaround to address comments.
